sdk/storage/Azure.Storage.Blobs/src/BlockBlobWriteStream.cs

```diff
@@ -0,0 +1,111 @@
+            IProgress<long> progressHandler) : base(
+                position,
+                bufferSize,
+                progressHandler)
+        {
+            ValidateBufferSize(bufferSize);
+            _blockBlobClient = blockBlobClient;
+            _conditions = conditions;
+            _blockIds = new List<string>();
+        }
+
+        protected override async Task AppendInternal(bool async, CancellationToken cancellationToken)
+        {
+            if (_buffer.Length > 0)
+            {
+                _buffer.Position = 0;
+
+                string blockId = GenerateBlockId();
+
+                if (async)
+                {
+                    await _blockBlobClient.StageBlockAsync(
+                        base64BlockId: blockId,
+                        content: _buffer,
+                        conditions: _conditions,
+                        progressHandler: _progressHandler,
+                        cancellationToken: cancellationToken)
+                        .ConfigureAwait(false);
+                }
+                else
+                {
+                    _blockBlobClient.StageBlock(
+                        base64BlockId: blockId,
+                        content: _buffer,
+                        conditions: _conditions,
+                        progressHandler: _progressHandler,
+                        cancellationToken: cancellationToken);
+                }
+
+                _blockIds.Add(blockId);
+                _buffer.Clear();
+            }
+        }
+
+        protected override async Task FlushInternal(bool async, CancellationToken cancellationToken)
+        {
+            await AppendInternal(async, cancellationToken).ConfigureAwait(false);
+
+            if (async)
+            {
+                await _blockBlobClient.CommitBlockListAsync(
+                    base64BlockIds: _blockIds,
+                    conditions: _conditions,
+                    cancellationToken: cancellationToken)
+                    .ConfigureAwait(false);
+            }
+            else
+            {
+                _blockBlobClient.CommitBlockList(
+                    base64BlockIds: _blockIds,
+                    conditions: _conditions,
+                    cancellationToken: cancellationToken);
+            }
+        }
+
+        protected override void ValidateBufferSize(long bufferSize)
+        {
+            if (bufferSize < 1)
+            {
+                throw new ArgumentOutOfRangeException(nameof(bufferSize), "Must be >= 1");
+            }
+
+            if (bufferSize > 100 * Constants.MB)
+            {
+                throw new ArgumentOutOfRangeException(nameof(bufferSize), "Must <= 100 MB");
+            }
+        }
+
+        private static string GenerateBlockId()
+        {
+            Guid guid = Guid.NewGuid();
+            byte[] bytes = Encoding.UTF8.GetBytes(guid.ToString());
+            return Convert.ToBase64String(bytes);
+        }
```

Comment:
Do the block IDs really all have to be the same length?  This seems like a strange requirement.

Reply:
Based on my previous comment about Put Blob not creating Blocks, I think we don't need to worry about this.

Reply:
Block IDs do have to be the same length. The service mandates it due to significant performance optimizations it can achieve with that constraint.

Leaving this discussion open until agreement is reached on how to deal with the block list situation.

Reply:
I've switched to using the same algorithm to generate Block IDs as partitioned upload.